File: pkg-r/R/cache.R
# Cache mode: offline existence checks against a pinned snapshot of valid ids.

.read_ids <- function(path) {
  ids <- trimws(readLines(path, warn = FALSE, encoding = "UTF-8"))
  ids[nzchar(ids)]
}

# Write lines to a path atomically: write a temporary file in the same directory,
# then rename it over the target in one step. A crash or a concurrent reader never
# sees a half-written file, so a truncated snapshot or cache cannot silently
# report valid ids as invalid. On Windows a rename cannot overwrite an existing
# file, so fall back to an overwriting copy there.
.atomic_write_lines <- function(lines, path) {
  tmp <- paste0(path, ".", Sys.getpid(), ".tmp")
  writeLines(lines, tmp)
  if (!isTRUE(file.rename(tmp, path))) {
    file.copy(tmp, path, overwrite = TRUE)
    unlink(tmp)
  }
  invisible(path)
}

.bundled_snapshots_dir <- function() {
  system.file("extdata", "snapshots", package = "biobouncer")
}

# Locate a snapshot file, plain or gzipped, user cache before bundled. suffix is
# ".txt" for the id set or ".retired.tsv" for the retired map. readLines() reads a
# .gz path transparently, so only the lookup needs to know about compression.
.find_snapshot <- function(source_db, version, suffix) {
  roots <- c(biobouncer_cache_dir(), .bundled_snapshots_dir())
  for (root in roots) {
    if (!nzchar(root)) {
      next
    }
    for (ext in c(suffix, paste0(suffix, ".gz"))) {
      path <- file.path(root, source_db, paste0(version, ext))
      if (file.exists(path)) {
        return(path)
      }
    }
  }
  NA_character_
}

.snapshot_file <- function(source_db, version) {
  .find_snapshot(source_db, version, ".txt")
}

.retired_file <- function(source_db, version) {
  .find_snapshot(source_db, version, ".retired.tsv")
}

.snapshot_versions <- function(source_db) {
  dirs <- c(
    file.path(biobouncer_cache_dir(), source_db),
    if (nzchar(.bundled_snapshots_dir())) {
      file.path(.bundled_snapshots_dir(), source_db)
    }
  )
  versions <- character(0)
  for (d in dirs) {
    if (dir.exists(d)) {
      versions <- c(
        versions,
        sub("\\.txt(\\.gz)?$", "", list.files(d, pattern = "\\.txt(\\.gz)?$"))
      )
    }
  }
  sort(unique(versions))
}

# The snapshot version cache mode uses when the caller gives none. Prefers the
# source's pinned default_version when a snapshot for it is installed, then the
# newest installed non-sample version, then a bundled sample. Returns NULL when
# nothing is installed. Newest is by sort order, which is chronological for the
# dated (ISO-8601) versions snapshots use. Mirrors default_cache_version() in the
# Python package.
.default_cache_version <- function(source_db, source) {
  installed <- .snapshot_versions(source_db)
  if (!length(installed)) {
    return(NULL)
  }
  dv <- source$default_version
  if (!is.null(dv) && dv %in% installed) {
    return(dv)
  }
  non_sample <- installed[installed != "sample"]
  if (length(non_sample)) {
    return(non_sample[length(non_sample)])
  }
  installed[length(installed)]
}

.snapshot_set <- function(source_db, version) {
  path <- .snapshot_file(source_db, version)
  if (is.na(path)) {
    available <- .snapshot_versions(source_db)
    cli::cli_abort(
      c(
        "No snapshot for {.val {source_db}} version {.val {version}}.",
        i = if (length(available)) {
          "Installed versions: {.val {available}}."
        } else {
          "No snapshots are installed for this source."
        },
        i = "Run {.code biobouncer_pull()} to download one."
      ),
      class = "biobouncer_error_missing_snapshot"
    )
  }
  .read_ids(path)
}

# Retired-id map for a snapshot version, read from the "<version>.retired.tsv"
# sidecar. Each non-blank line is "retired<TAB>successor"; the successor is the
# second tab field, or "" when absent. Returns a named character vector mapping
# retired id to successor, empty when the source has no sidecar.
.snapshot_retired <- function(source_db, version) {
  path <- .retired_file(source_db, version)
  empty <- character(0)
  names(empty) <- character(0)
  if (is.na(path)) {
    return(empty)
  }
  lines <- trimws(readLines(path, warn = FALSE, encoding = "UTF-8"))
  lines <- lines[nzchar(lines)]
  if (!length(lines)) {
    return(empty)
  }
  keys <- character(length(lines))
  successors <- character(length(lines))
  for (i in seq_along(lines)) {
    fields <- strsplit(lines[i], "\t", fixed = TRUE)[[1]]
    keys[i] <- fields[1]
    successors[i] <- if (length(fields) >= 2L) fields[2] else ""
  }
  names(successors) <- keys
  successors
}

.cache_verdicts <- function(source, x, is_na, ids, retired = character(0)) {
  n <- length(x)
  valid <- rep(NA, n)
  normalized <- rep(NA_character_, n)
  suggestion <- rep(NA_character_, n)
  wellformed <- rep(NA, n)
  wellformed[!is_na] <- .matches(source$pattern, x[!is_na])
  for (i in seq_len(n)) {
    if (is_na[i]) {
      next
    }
    if (isTRUE(wellformed[i])) {
      if (x[i] %in% ids) {
        valid[i] <- TRUE
        normalized[i] <- x[i]
      } else {
        valid[i] <- FALSE
        if (x[i] %in% names(retired)) {
          succ <- retired[[x[i]]]
          if (nzchar(succ)) {
            suggestion[i] <- succ
          }
        }
      }
    } else {
      valid[i] <- FALSE
      sugg <- .suggest_one(source, x[i])
      if (!is.na(sugg) && sugg %in% ids) {
        suggestion[i] <- sugg
      }
    }
    if (isFALSE(valid[i]) && is.na(suggestion[i])) {
      fz <- .fuzzy_suggest(source, x[i], ids)
      if (!is.na(fz)) {
        suggestion[i] <- fz
      }
    }
  }
  list(valid = valid, normalized = normalized, suggestion = suggestion)
}

#' Snapshot cache directory
#'
#' The directory where downloaded snapshots are stored. Set the environment
#' variable `BIOBOUNCER_CACHE_DIR` to override the default.
#'
#' @return A path to the cache directory.
#' @examples
#' biobouncer_cache_dir()
#' @export
biobouncer_cache_dir <- function() {
  override <- Sys.getenv("BIOBOUNCER_CACHE_DIR", unset = "")
  if (nzchar(override)) {
    return(override)
  }
  tools::R_user_dir("biobouncer", which = "cache")
}

#' List installed snapshots
#'
#' Reports snapshots available for `cache` mode, both downloaded ones in the
#' cache directory and the small bundled samples.
#'
#' @return A [tibble][tibble::tibble] with columns `source`, `version`,
#'   `n_ids`, and `location` (`"cache"` or `"bundled"`).
#' @examples
#' biobouncer_snapshots()
#' @export
biobouncer_snapshots <- function() {
  locations <- list(
    cache = biobouncer_cache_dir(),
    bundled = .bundled_snapshots_dir()
  )
  rows <- list()
  for (loc in names(locations)) {
    base <- locations[[loc]]
    if (!nzchar(base) || !dir.exists(base)) {
      next
    }
    files <- list.files(
      base,
      pattern = "\\.txt(\\.gz)?$",
      recursive = TRUE,
      full.names = TRUE
    )
    for (path in files) {
      rows[[length(rows) + 1L]] <- data.frame(
        source = basename(dirname(path)),
        version = sub("\\.txt(\\.gz)?$", "", basename(path)),
        n_ids = length(.read_ids(path)),
        location = loc,
        stringsAsFactors = FALSE
      )
    }
  }
  if (!length(rows)) {
    return(tibble::tibble(
      source = character(0),
      version = character(0),
      n_ids = integer(0),
      location = character(0)
    ))
  }
  tibble::as_tibble(do.call(rbind, rows))
}

.sanitize_version <- function(raw) {
  raw <- sub("^releases/", "", trimws(raw))
  gsub("[^A-Za-z0-9._-]", "-", raw)
}

# Extract (version, ids) from OBO lines, keeping ids that match the pattern.
.parse_obo <- function(lines, pattern) {
  version <- NA_character_
  dv <- grep("^data-version:", lines, value = TRUE)
  if (length(dv)) {
    version <- .sanitize_version(sub("^data-version:", "", dv[1]))
  }
  id_values <- sub("^id:\\s*", "", grep("^id:\\s", lines, value = TRUE))
  ids <- sort(unique(id_values[.matches(pattern, id_values)]), method = "radix")
  list(
    version = if (is.na(version) || !nzchar(version)) NULL else version,
    ids = ids
  )
}

# Empty retired map: a named character vector with no entries.
.empty_retired <- function() {
  out <- character(0)
  names(out) <- character(0)
  out
}

# Split an HGNC multi-value field on "|", trimming surrounding quotes and spaces.
.split_pipe <- function(field) {
  field <- gsub('^"|"$', "", trimws(field))
  if (!nzchar(field)) {
    return(character(0))
  }
  toks <- gsub('^"|"$', "", trimws(strsplit(field, "|", fixed = TRUE)[[1]]))
  toks[nzchar(toks)]
}

# Extract (version, approved ids, retired map) from an HGNC complete-set TSV.
# Mirrors parse_hgnc_tsv() in the Python package exactly: the approved set is
# every "symbol" with status "Approved" that matches the pattern; the retired
# map sends a previous or alias symbol to its approved successor, a previous
# symbol wins over an alias, an approved symbol is never retired, and an
# ambiguous mapping is dropped. Everything is sorted by code point (radix) so
# the two languages write byte-identical snapshots.
.parse_hgnc_tsv <- function(lines, pattern) {
  if (!length(lines)) {
    return(list(version = NULL, ids = character(0), retired = .empty_retired()))
  }
  header <- strsplit(lines[1], "\t", fixed = TRUE)[[1]]
  c_symbol <- match("symbol", header)
  c_status <- match("status", header)
  if (is.na(c_symbol) || is.na(c_status)) {
    cli::cli_abort(
      "HGNC TSV is missing the {.field symbol} or {.field status} column."
    )
  }
  c_prev <- match("prev_symbol", header)
  c_alias <- match("alias_symbol", header)

  data <- lines[-1]
  data <- data[nzchar(trimws(data))]
  fields_list <- strsplit(data, "\t", fixed = TRUE)
  col <- function(i) {
    vapply(
      fields_list,
      function(f) if (length(f) >= i) f[i] else NA_character_,
      character(1)
    )
  }
  symbol <- trimws(col(c_symbol))
  status <- trimws(col(c_status))
  ok <- !is.na(symbol) & nzchar(symbol) & !is.na(status) & status == "Approved"
  ok[ok] <- .matches(pattern, symbol[ok])
  approved <- sort(unique(symbol[ok]), method = "radix")

  prev_map <- list()
  alias_map <- list()
  for (j in which(ok)) {
    sym <- symbol[j]
    fields <- fields_list[[j]]
    if (!is.na(c_prev) && length(fields) >= c_prev) {
      for (old in .split_pipe(fields[c_prev])) {
        prev_map[[old]] <- unique(c(prev_map[[old]], sym))
      }
    }
    if (!is.na(c_alias) && length(fields) >= c_alias) {
      for (old in .split_pipe(fields[c_alias])) {
        alias_map[[old]] <- unique(c(alias_map[[old]], sym))
      }
    }
  }

  keys <- sort(unique(c(names(prev_map), names(alias_map))), method = "radix")
  keys <- keys[nzchar(keys) & !(keys %in% approved)]
  if (length(keys)) {
    keys <- keys[.matches(pattern, keys)]
  }
  retired_keys <- character(0)
  retired_vals <- character(0)
  for (old in keys) {
    targets <- if (!is.null(prev_map[[old]])) {
      prev_map[[old]]
    } else {
      alias_map[[old]]
    }
    if (length(targets) == 1L) {
      retired_keys <- c(retired_keys, old)
      retired_vals <- c(retired_vals, targets)
    }
  }
  names(retired_vals) <- retired_keys
  list(version = NULL, ids = approved, retired = retired_vals)
}

# Snapshot builders, keyed by the source's cache$builder. Each has
# url(source, version) and build(lines, source) -> list(version, ids, retired).
# Mirrors the Python _BUILDERS registry.
.builders <- list(
  obo = list(
    url = function(source, version) source$cache$obo_url,
    build = function(lines, source) {
      parsed <- .parse_obo(lines, source$pattern)
      list(
        version = parsed$version,
        ids = parsed$ids,
        retired = .empty_retired()
      )
    }
  ),
  hgnc_tsv = list(
    url = function(source, version) {
      template <- source$cache$tsv_url
      resolved <- if (!is.null(version)) version else source$default_version
      if (grepl("{version}", template, fixed = TRUE)) {
        gsub("{version}", resolved, template, fixed = TRUE)
      } else {
        template
      }
    },
    build = function(lines, source) .parse_hgnc_tsv(lines, source$pattern)
  )
)

#' Download a snapshot for cache mode
#'
#' Dispatches on the source's `cache$builder`: `obo` fetches the ontology
#' release, `hgnc_tsv` fetches the HGNC complete set. Identifiers that match the
#' source pattern are written to the cache directory as a snapshot, and a
#' retired-id map, when the builder produces one, to the matching
#' `<version>.retired.tsv` sidecar. An OBO version defaults to the ontology's own
#' data-version; an HGNC version defaults to the source's `default_version`.
#'
#' @param source_db Source key, for example `"mondo"`.
#' @param version Snapshot version label. Defaults to the builder's own version.
#' @param quiet Suppress progress messages.
#' @return The path to the written snapshot, invisibly.
#' @seealso [biobouncer_snapshots()], [check_id()].
#' @export
biobouncer_pull <- function(source_db, version = NULL, quiet = FALSE) {
  source <- .get_source(source_db)
  cache <- source$cache
  builder <- if (is.null(cache) || is.null(cache$builder)) {
    NULL
  } else {
    .builders[[cache$builder]]
  }
  if (is.null(builder)) {
    cli::cli_abort(
      "No snapshot builder is available for {.val {source_db}}.",
      class = "biobouncer_error_no_builder"
    )
  }
  url <- builder$url(source, version)
  tmp <- tempfile()
  on.exit(unlink(tmp), add = TRUE)
  if (!quiet) {
    cli::cli_inform("Downloading {.url {url}} ...")
  }
  utils::download.file(
    url,
    tmp,
    quiet = quiet,
    mode = "wb",
    headers = c(
      "User-Agent" = "biobouncer/0.1 (+https://github.com/samuelbharti/biobouncer)"
    )
  )
  built <- builder$build(
    readLines(tmp, warn = FALSE, encoding = "UTF-8"),
    source
  )
  version <- if (!is.null(version)) {
    as.character(version)
  } else if (!is.null(built$version)) {
    built$version
  } else {
    source$default_version
  }
  if (is.null(version) || !nzchar(version)) {
    cli::cli_abort(
      "Could not determine a version for {.val {source_db}}; pass {.arg version}.",
      class = "biobouncer_error_missing_version"
    )
  }
  dest_dir <- file.path(biobouncer_cache_dir(), source_db)
  dir.create(dest_dir, recursive = TRUE, showWarnings = FALSE)
  dest <- file.path(dest_dir, paste0(version, ".txt"))
  .atomic_write_lines(built$ids, dest)
  retired <- built$retired
  if (length(retired)) {
    keys <- sort(names(retired), method = "radix")
    .atomic_write_lines(
      paste(keys, retired[keys], sep = "\t"),
      file.path(dest_dir, paste0(version, ".retired.tsv"))
    )
  }
  if (!quiet) {
    cli::cli_inform("Wrote {length(built$ids)} ids to {.file {dest}}.")
  }
  invisible(dest)
}

File: pkg-py/src/biogate/_cache.py
"""Cache mode: offline existence checks against a pinned snapshot of valid ids."""

from __future__ import annotations

import gzip
import os
import re
import urllib.request
from collections.abc import Callable
from dataclasses import dataclass
from importlib.resources import files
from pathlib import Path

import platformdirs

from ._fuzzy import fuzzy_suggest
from ._pattern import _suggest, matches
from ._registry import Source, get_source


class MissingVersionError(ValueError):
    """Raised when cache mode is used without a version."""


class MissingSnapshotError(FileNotFoundError):
    """Raised when no snapshot is installed for a (source, version)."""


def cache_dir() -> Path:
    """Directory where downloaded snapshots are stored.

    Set the ``BIOGATE_CACHE_DIR`` environment variable to override the default.
    """
    override = os.environ.get("BIOGATE_CACHE_DIR")
    if override:
        return Path(override)
    return Path(platformdirs.user_cache_dir("biogate"))


def _atomic_write_text(path: Path, text: str, encoding: str = "utf-8") -> None:
    """Write ``text`` to ``path`` atomically.

    Write to a temporary file in the same directory, then replace the target in
    one step with ``os.replace``. A crash or a concurrent reader never sees a
    half-written file, so a truncated snapshot or cache cannot silently report
    valid ids as invalid.
    """
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_name(f"{path.name}.{os.getpid()}.tmp")
    tmp.write_text(text, encoding=encoding)
    os.replace(tmp, path)


def _bundled_root():
    return files("biogate") / "_data" / "snapshots"


def _ids_from_text(text: str) -> set[str]:
    return {line.strip() for line in text.splitlines() if line.strip()}


def _read_snapshot(res) -> str:
    """Read a snapshot resource, transparently decompressing a ``.gz`` file."""
    data = res.read_bytes()
    if res.name.endswith(".gz"):
        data = gzip.decompress(data)
    return data.decode("utf-8")


def _find_snapshot(source_db: str, version: str, suffix: str):
    """Locate a snapshot file, plain or gzipped, user cache before bundled.

    ``suffix`` is ``".txt"`` for the id set or ``".retired.tsv"`` for the retired
    map. Returns a path-like (a filesystem ``Path`` or a package resource) or
    ``None`` when neither form is installed.
    """
    user = cache_dir() / source_db
    for ext in (suffix, suffix + ".gz"):
        candidate = user / f"{version}{ext}"
        if candidate.is_file():
            return candidate
    bundled = _bundled_root() / source_db
    for ext in (suffix, suffix + ".gz"):
        candidate = bundled / f"{version}{ext}"
        if candidate.is_file():
            return candidate
    return None


def _strip_snapshot_ext(name: str) -> str | None:
    """Return the version for an id-snapshot filename, or ``None`` if it is not one."""
    if name.endswith(".txt.gz"):
        return name[: -len(".txt.gz")]
    if name.endswith(".txt"):
        return name[: -len(".txt")]
    return None


def _snapshot_text(source_db: str, version: str) -> str | None:
    res = _find_snapshot(source_db, version, ".txt")
    return _read_snapshot(res) if res is not None else None


def _retired_text(source_db: str, version: str) -> str | None:
    res = _find_snapshot(source_db, version, ".retired.tsv")
    return _read_snapshot(res) if res is not None else None


def _snapshot_retired(source_db: str, version: str) -> dict[str, str]:
    """Map retired id to successor from the ``<version>.retired.tsv`` sidecar.

    Each non-blank line is a retired id and its successor separated by a tab. A
    line with no successor maps to an empty string. No sidecar yields an empty
    map.
    """
    text = _retired_text(source_db, version)
    if text is None:
        return {}
    mapping: dict[str, str] = {}
    for line in text.splitlines():
        if not line.strip():
            continue
        fields = line.split("\t")
        mapping[fields[0]] = fields[1] if len(fields) > 1 else ""
    return mapping


def _snapshot_versions(source_db: str) -> list[str]:
    versions: set[str] = set()
    user = cache_dir() / source_db
    if user.is_dir():
        for p in user.iterdir():
            version = _strip_snapshot_ext(p.name)
            if version is not None:
                versions.add(version)
    bundled = _bundled_root() / source_db
    if bundled.is_dir():
        for c in bundled.iterdir():
            version = _strip_snapshot_ext(c.name)
            if version is not None:
                versions.add(version)
    return sorted(versions)


def snapshot_set(source_db: str, version: str) -> set[str]:
    text = _snapshot_text(source_db, version)
    if text is None:
        available = _snapshot_versions(source_db)
        detail = (
            f"Installed versions: {', '.join(available)}."
            if available
            else "No snapshots are installed for this source."
        )
        raise MissingSnapshotError(
            f"No snapshot for {source_db!r} version {version!r}. {detail} "
            "Run biogate.pull() to download one."
        )
    return _ids_from_text(text)


def _fuzzy_fallback(
    s: str, fuzzy: tuple[dict[int, list[str]], int] | None
) -> str | None:
    if fuzzy is None:
        return None
    index, max_distance = fuzzy
    return fuzzy_suggest(s, index, max_distance)


def cache_check(
    source: Source,
    s: str,
    ids: set[str],
    retired: dict[str, str] | None = None,
    fuzzy: tuple[dict[int, list[str]], int] | None = None,
) -> tuple[bool, str | None, str | None]:
    """Return ``(valid, normalized, suggestion)`` for a single input.

    A well-formed id absent from ``ids`` but present in ``retired`` with a
    non-empty successor is invalid and suggests that successor. ``fuzzy`` is an
    optional ``(length index, max distance)`` pair; when set, an id with no exact,
    retired, or normalized match falls back to the nearest id within that edit
    distance.
    """
    retired = retired or {}
    if matches(source.pattern, s):
        if s in ids:
            return True, s, None
        successor = retired.get(s)
        if successor:
            return False, None, successor
        return False, None, _fuzzy_fallback(s, fuzzy)
    suggestion = _suggest(source, s)
    if suggestion is not None and suggestion in ids:
        return False, None, suggestion
    return False, None, _fuzzy_fallback(s, fuzzy)


def _snapshot_rows(base, location: str) -> list[dict]:
    rows: list[dict] = []
    if not base.is_dir():
        return rows
    for src_dir in sorted(base.iterdir(), key=lambda c: c.name):
        if not src_dir.is_dir():
            continue
        for f in sorted(src_dir.iterdir(), key=lambda c: c.name):
            version = _strip_snapshot_ext(f.name)
            if version is None:
                continue
            rows.append(
                {
                    "source": src_dir.name,
                    "version": version,
                    "n_ids": len(_ids_from_text(_read_snapshot(f))),
                    "location": location,
                }
            )
    return rows


def snapshots() -> list[dict]:
    """List installed snapshots, both cached and bundled."""
    return _snapshot_rows(cache_dir(), "cache") + _snapshot_rows(
        _bundled_root(), "bundled"
    )


_USER_AGENT = "biogate/0.1 (+https://github.com/samuelbharti/biogate)"


class NoBuilderError(ValueError):
    """Raised when a source has no snapshot builder."""


def _sanitize_version(raw: str) -> str:
    raw = raw.strip().removeprefix("releases/")
    return re.sub(r"[^A-Za-z0-9._-]", "-", raw)


def parse_obo(text: str, pattern: str) -> tuple[str | None, list[str]]:
    """Extract (version, ids) from OBO text, keeping ids that match the pattern."""
    version: str | None = None
    ids: set[str] = set()
    for line in text.splitlines():
        if version is None and line.startswith("data-version:"):
            version = _sanitize_version(line[len("data-version:") :]) or None
        elif line.startswith("id:"):
            value = line[len("id:") :].strip()
            if matches(pattern, value):
                ids.add(value)
    return version, sorted(ids)


def _split_pipe(field: str) -> list[str]:
    """Split an HGNC multi-value field on ``|``, trimming quotes and spaces."""
    field = field.strip().strip('"')
    if not field:
        return []
    return [tok.strip().strip('"') for tok in field.split("|") if tok.strip()]


def parse_hgnc_tsv(
    text: str, pattern: str
) -> tuple[str | None, list[str], dict[str, str]]:
    """Extract (version, approved ids, retired map) from an HGNC complete-set TSV.

    Columns are looked up by name, never by position. The approved set is every
    ``symbol`` whose ``status`` is ``Approved`` and matches the source pattern.
    The retired map sends a previous or alias symbol to its approved successor:
    a previous symbol wins over an alias, a symbol that is itself approved is
    never retired, and an ambiguous mapping (two or more approved successors) is
    dropped. Ids and retired keys are sorted by Unicode code point so R and
    Python write byte-identical snapshots. The version is always ``None`` here
    (an HGNC snapshot is dated by its download, not by its content).
    """
    lines = text.splitlines()
    if not lines:
        return None, [], {}
    header = lines[0].split("\t")
    index = {name: i for i, name in enumerate(header)}
    if "symbol" not in index or "status" not in index:
        raise ValueError("HGNC TSV is missing the 'symbol' or 'status' column.")
    c_symbol, c_status = index["symbol"], index["status"]
    c_prev, c_alias = index.get("prev_symbol"), index.get("alias_symbol")

    approved: set[str] = set()
    prev_map: dict[str, set[str]] = {}
    alias_map: dict[str, set[str]] = {}
    for line in lines[1:]:
        if not line.strip():
            continue
        fields = line.split("\t")
        if len(fields) <= max(c_symbol, c_status):
            continue
        if fields[c_status].strip() != "Approved":
            continue
        symbol = fields[c_symbol].strip()
        if not symbol or not matches(pattern, symbol):
            continue
        approved.add(symbol)
        if c_prev is not None and c_prev < len(fields):
            for old in _split_pipe(fields[c_prev]):
                prev_map.setdefault(old, set()).add(symbol)
        if c_alias is not None and c_alias < len(fields):
            for old in _split_pipe(fields[c_alias]):
                alias_map.setdefault(old, set()).add(symbol)

    retired: dict[str, str] = {}
    for old in sorted(set(prev_map) | set(alias_map)):
        if not old or old in approved or not matches(pattern, old):
            continue
        targets = prev_map.get(old) or alias_map.get(old)
        if targets and len(targets) == 1:
            retired[old] = next(iter(targets))
    return None, sorted(approved), retired


@dataclass(frozen=True)
class _Builder:
    url: Callable[[Source, str | None], str]
    build: Callable[[str, Source], tuple[str | None, list[str], dict[str, str]]]


def _obo_url(source: Source, version: str | None) -> str:
    return source.cache["obo_url"]


def _obo_build(
    text: str, source: Source
) -> tuple[str | None, list[str], dict[str, str]]:
    version, ids = parse_obo(text, source.pattern)
    return version, ids, {}


def _hgnc_url(source: Source, version: str | None) -> str:
    template = source.cache["tsv_url"]
    resolved = version if version is not None else source.default_version
    return template.format(version=resolved) if "{version}" in template else template


def _hgnc_build(
    text: str, source: Source
) -> tuple[str | None, list[str], dict[str, str]]:
    return parse_hgnc_tsv(text, source.pattern)


_BUILDERS: dict[str, _Builder] = {
    "obo": _Builder(_obo_url, _obo_build),
    "hgnc_tsv": _Builder(_hgnc_url, _hgnc_build),
}


def pull(
    source_db: str, version: str | None = None, quiet: bool = False, timeout: int = 120
) -> Path:
    """Download a full snapshot for cache mode into the cache directory.

    Dispatches on the source's ``cache.builder``: ``obo`` fetches the ontology
    release, ``hgnc_tsv`` fetches the HGNC complete set. Identifiers that match
    the source pattern are written to ``cache_dir()/<source>/<version>.txt``, and
    a retired-id map, when the builder produces one, to the matching
    ``<version>.retired.tsv`` sidecar. An OBO version defaults to the ontology's
    own data-version; an HGNC version defaults to the source's ``default_version``.
    """
    source = get_source(source_db)
    cache = source.cache
    builder = _BUILDERS.get(cache.get("builder")) if cache else None
    if builder is None:
        raise NoBuilderError(f"No snapshot builder is available for {source_db!r}.")
    url = builder.url(source, version)
    if not quiet:
        print(f"Downloading {url} ...")
    request = urllib.request.Request(url, headers={"User-Agent": _USER_AGENT})
    with urllib.request.urlopen(request, timeout=timeout) as response:  # noqa: S310
        text = response.read().decode("utf-8", errors="replace")
    parsed_version, ids, retired = builder.build(text, source)
    version = (
        str(version)
        if version is not None
        else (parsed_version or source.default_version)
    )
    if not version:
        raise MissingVersionError(
            f"Could not determine a version for {source_db!r}; pass version."
        )
    dest = cache_dir() / source_db / f"{version}.txt"
    _atomic_write_text(dest, "\n".join(ids) + "\n")
    if retired:
        lines = [f"{old}\t{succ}" for old, succ in sorted(retired.items())]
        _atomic_write_text(
            cache_dir() / source_db / f"{version}.retired.tsv", "\n".join(lines) + "\n"
        )
    if not quiet:
        print(f"Wrote {len(ids)} ids to {dest}")
    return dest
